# loc, iloc and groupby

In [1]:
import pandas as pd # initializing panda

In [2]:
df = pd.read_csv("tips.csv") # reading .csv file using panda

In [3]:
df # print csv file

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [4]:
type(df)

pandas.core.frame.DataFrame

In [5]:
df.head() # previews first 5 rows of a DataFrame

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [6]:
df.tail(3) # previews last 3 rows of a DataFrame; by default  5 ota garxa but we can change that

total_bill   tip     sex smoker   day    time  size
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [7]:
df.sample(5) # randomly samples 5 rows

total_bill   tip     sex smoker   day    time  size
158       13.39  2.61  Female     No   Sun  Dinner     2
134       18.26  3.25  Female     No  Thur   Lunch     2
215       12.90  1.10  Female    Yes   Sat  Dinner     2
28        21.70  4.30    Male     No   Sat  Dinner     2
14        14.83  3.02  Female     No   Sun  Dinner     2

In [8]:
# yo sampel harek patak run garda naya auxa  same aunw banuana seed chainxa
# pandas ra scikit-learn ma seed lai random_state vanxa

In [9]:
df.sample(5, random_state=2024)

total_bill    tip     sex smoker   day    time  size
239       29.03   5.92    Male     No   Sat  Dinner     3
238       35.83   4.67  Female     No   Sat  Dinner     3
170       50.81  10.00    Male    Yes   Sat  Dinner     3
156       48.17   5.00    Male     No   Sun  Dinner     6
134       18.26   3.25  Female     No  Thur   Lunch     2

In [10]:
df.shape # (no_of_rows, no_of_column)

(244, 7)

# Data profiling and inspection

In [11]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

In [12]:
df.head(2)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3

# Data type in csv are numerical, categorical and numerical-categorical

In [13]:
df.describe() # describes the numerical data

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

In [14]:
# 25% -> Q1
# 50% -> Q2/median
# 75% -> Q3

In [15]:
df.dtypes

total_bill    float64
tip           float64
sex            object
smoker         object
day            object
time           object
size            int64
dtype: object

In [16]:
df.describe(include='object')

sex smoker  day    time
count    244    244  244     244
unique     2      2    4       2
top     Male     No  Sat  Dinner
freq     157    151   87     176

# Sex column
- 244 rows
- 2 unique categories
- among 2 unique categories most occuring is Male
- with 157 frequency

In [17]:
df.sex.unique()

array(['Female', 'Male'], dtype=object)

In [18]:
df.sex.nunique()

2

In [19]:
df.sex.value_counts()

sex
Male      157
Female     87
Name: count, dtype: int64

# Classwork time and day column

In [20]:
# time
df.time.unique()

array(['Dinner', 'Lunch'], dtype=object)

In [21]:
df['time'].nunique()

2

In [22]:
df["time"].value_counts()

time
Dinner    176
Lunch      68
Name: count, dtype: int64

In [23]:
# day
df['day'].unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [24]:
df['day'].nunique()

4

In [25]:
df['day'].value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


# NULL means missing data

In [27]:
df.isna() # checks if the data contanis null or not

total_bill    tip    sex  smoker    day   time   size
0         False  False  False   False  False  False  False
1         False  False  False   False  False  False  False
2         False  False  False   False  False  False  False
3         False  False  False   False  False  False  False
4         False  False  False   False  False  False  False
..          ...    ...    ...     ...    ...    ...    ...
239       False  False  False   False  False  False  False
240       False  False  False   False  False  False  False
241       False  False  False   False  False  False  False
242       False  False  False   False  False  False  False
243       False  False  False   False  False  False  False

[244 rows x 7 columns]

In [28]:
# I want to count no of missing values ini each column

In [29]:
True + True

2

In [30]:
True + False

1

In [31]:
False + True

1

In [32]:
False + False

0

In [33]:
df.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [34]:
# no missing data

# loc VS iloc

In [35]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

# Slice data  with Male in sex column

In [36]:
df['sex'] == "Male"

0      False
1       True
2       True
3       True
4      False
       ...  
239     True
240    False
241     True
242     True
243    False
Name: sex, Length: 244, dtype: bool

In [37]:
df.loc[df['sex']=='Male'] # slicing data with Male in sex column

total_bill   tip   sex smoker  day    time  size
1         10.34  1.66  Male     No  Sun  Dinner     3
2         21.01  3.50  Male     No  Sun  Dinner     3
3         23.68  3.31  Male     No  Sun  Dinner     2
5         25.29  4.71  Male     No  Sun  Dinner     4
6          8.77  2.00  Male     No  Sun  Dinner     2
..          ...   ...   ...    ...  ...     ...   ...
236       12.60  1.00  Male    Yes  Sat  Dinner     2
237       32.83  1.17  Male    Yes  Sat  Dinner     2
239       29.03  5.92  Male     No  Sat  Dinner     3
241       22.67  2.00  Male    Yes  Sat  Dinner     2
242       17.82  1.75  Male     No  Sat  Dinner     2

[157 rows x 7 columns]

# Slice data  with Female in sex column

In [38]:
df['sex'] == "Female"

0       True
1      False
2      False
3      False
4       True
       ...  
239    False
240     True
241    False
242    False
243     True
Name: sex, Length: 244, dtype: bool

In [39]:
df.loc[df['sex']=='Female']

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
11        35.26  5.00  Female     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
..          ...   ...     ...    ...   ...     ...   ...
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[87 rows x 7 columns]

# Who gives more tips on average Male Or Female?

In [40]:
df.loc[df['sex']=="Male"]

total_bill   tip   sex smoker  day    time  size
1         10.34  1.66  Male     No  Sun  Dinner     3
2         21.01  3.50  Male     No  Sun  Dinner     3
3         23.68  3.31  Male     No  Sun  Dinner     2
5         25.29  4.71  Male     No  Sun  Dinner     4
6          8.77  2.00  Male     No  Sun  Dinner     2
..          ...   ...   ...    ...  ...     ...   ...
236       12.60  1.00  Male    Yes  Sat  Dinner     2
237       32.83  1.17  Male    Yes  Sat  Dinner     2
239       29.03  5.92  Male     No  Sat  Dinner     3
241       22.67  2.00  Male    Yes  Sat  Dinner     2
242       17.82  1.75  Male     No  Sat  Dinner     2

[157 rows x 7 columns]

In [41]:
df.loc[df['sex']=="Male"]["tip"]


1      1.66
2      3.50
3      3.31
5      4.71
6      2.00
       ... 
236    1.00
237    1.17
239    5.92
241    2.00
242    1.75
Name: tip, Length: 157, dtype: float64

In [42]:
df.loc[df['sex']=="Male"]["tip"].mean()

3.0896178343949043

In [43]:
df.loc[df['sex']=='Female']

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
11        35.26  5.00  Female     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
..          ...   ...     ...    ...   ...     ...   ...
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[87 rows x 7 columns]

In [44]:
df.loc[df['sex']=='Female']['tip']

0      1.01
4      3.61
11     5.00
14     3.02
16     1.67
       ... 
226    2.00
229    2.88
238    4.67
240    2.00
243    3.00
Name: tip, Length: 87, dtype: float64

In [45]:
df.loc[df['sex']=='Female']['tip'].mean()

2.8334482758620685

In [46]:
df.day.unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [47]:
# kun bar on average kati tip hunxa; sab vanda badi tip kun bar dinxa?

In [48]:
df.loc[df['day'] == 'Sun']['tip'].mean()

3.2551315789473687

In [49]:
df.loc[df['day']== 'Sat']['tip'].mean()


2.993103448275862

In [50]:
# yo ta each category lai garnu paryo; if categories are many its difficult to do this

In [51]:
# esko solution groupby ho

# Groupby

In [52]:
# day ko adhar ma data lai group garnu paryo

In [53]:
df.groupby(['day'])

In [54]:
df.groupby(['day'])['tip'].mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [55]:
# or
df.groupby(['day']).tip.mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [56]:
df.groupby(['sex','smoker']).tip.mean()

sex     smoker
Female  No        2.773519
        Yes       2.931515
Male    No        3.113402
        Yes       3.051167
Name: tip, dtype: float64

# loc VS iloc

In [57]:
# loc column slice garnu bhaki xa

# i want rows 14, 15, 16, and all columns

In [58]:
df.loc[14:16,:]

total_bill   tip     sex smoker  day    time  size
14       14.83  3.02  Female     No  Sun  Dinner     2
15       21.58  3.92    Male     No  Sun  Dinner     2
16       10.33  1.67  Female     No  Sun  Dinner     3

In [59]:
# loc use  garda row slice garda last ko index linxa eg: 14:16 means 14, 15, 16

# Row 14-16 and column tip. sex

In [60]:
df.loc[14:16, ['tip','sex']]

tip     sex
14  3.02  Female
15  3.92    Male
16  1.67  Female

In [61]:
# kasaile tip ra sex column ta index 1 and 3 ma xa why not to use index instead of column name
df.loc[14:16, 1:2]



TypeError: cannot do slice indexing on Index with these indexers [1] of type int

In [62]:
 # loc le column name matra accept garxa; column ko index gardaina

In [ ]:
# Yedi column ko index diney ho vane iloc use garnu parxa

In [ ]:
df.iloc[14:16, 1:3]

In [ ]:
# iloc le row ra column slice garda last ko index lidaina

In [ ]:
# iloc ma column ko index diney ho name dina paudaina

# Slicing DataFrame

In [ ]:
# i want sabai rows tip and total_bill column

In [ ]:
df[['tip','total_bill']]

In [ ]:
df.loc[:, ['tip','total_bill']]